src/Components/Browser.JS/src/Boot.Server.ts

```diff
@@ -6,70 +6,97 @@ import { OutOfProcessRenderBatch } from './Rendering/RenderBatch/OutOfProcessRen
 import { fetchBootConfigAsync, loadEmbeddedResourcesAsync } from './BootCommon';
+import { CircuitHandler } from './Platform/Circuits/CircuitHandler';
 
-let connection : signalR.HubConnection;
+async function boot(reconnect: boolean = false) {
+  const circuitHandlers = new Array<CircuitHandler>();
+  window['Blazor'].addCircuitHandler = (circuitHandler: CircuitHandler) => circuitHandlers.push(circuitHandler);
 
-function boot() {
-  // In the background, start loading the boot config and any embedded resources
-  const embeddedResourcesPromise = fetchBootConfigAsync().then(bootConfig => {
-    return loadEmbeddedResourcesAsync(bootConfig);
-  });
+  await startCicuit();
 
-  connection = new signalR.HubConnectionBuilder()
-    .withUrl('_blazor')
-    .withHubProtocol(new MessagePackHubProtocol())
-    .configureLogging(signalR.LogLevel.Information)
-    .build();
+  async function startCicuit(): Promise<void> {
+    // In the background, start loading the boot config and any embedded resources
+    const embeddedResourcesPromise = fetchBootConfigAsync().then(bootConfig => {
+      return loadEmbeddedResourcesAsync(bootConfig);
+    });
+
+    const initialConnection = await initializeConnection();
+
+    // Ensure any embedded resources have been loaded before starting the app
+    await embeddedResourcesPromise;
+    const circuitId = await initialConnection.invoke<string>(
+      'StartCircuit',
+      uriHelperFunctions.getLocationHref(),
+      uriHelperFunctions.getBaseURI()
+    );
+
+    window['Blazor'].reconnect = async () => {
+      const reconnection = await initializeConnection();
+      if (!await reconnection.invoke<Boolean>('ConnectCircuit', circuitId)) {
+        throw "Failed to reconnect to the server";
+      }
+
+      circuitHandlers.forEach(h => h.onConnectionUp());
+    };
+
+    circuitHandlers.forEach(h => h.onConnectionUp());
+  }
+
+  async function initializeConnection(): Promise<signalR.HubConnection> {
+    const connection = new signalR.HubConnectionBuilder()
+      .withUrl('_blazor')
+      .withHubProtocol(new MessagePackHubProtocol())
+      .configureLogging(signalR.LogLevel.Information)
+      .build();
+    connection.on('JS.BeginInvokeJS', DotNet.jsCallDispatcher.beginInvokeJSFromDotNet);
+    connection.on('JS.RenderBatch', (browserRendererId: number, renderId: number, batchData: Uint8Array) => {
+      try {
+        renderBatch(browserRendererId, new OutOfProcessRenderBatch(batchData));
+        connection.send('OnRenderCompleted', renderId, null);
+      }
+      catch (ex) {
+        // If there's a rendering exception, notify server *and* throw on client
+        connection.send('OnRenderCompleted', renderId, ex.toString());
+        throw ex;
+      }
+    });
+
+    connection.onclose(error => circuitHandlers.forEach(h => h.onConnectionDown(error)));
+    connection.on('JS.Error', unhandledError.bind(connection));
+
+    window['Blazor'].closeConnection = async() => {
```

Comment:
Do we definitely need to expose an API for this? Perhaps you're only using it for testing, but I don't see any obvious reason it should be part of the official API.

Reply:
It's meant for testing. There wasn't any other way to test out a websocket close locally since offline mode in developer tools had no affect on websockets.